src/theory/symmetry_detect.cpp

```diff
@@ -0,0 +1,525 @@
+/*********************                                                        */
+/*! \file symmetry_detect.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **  Paul Meng, Andrew Reynolds
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2018 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Symmetry detection module
+ **/
+
+#include "theory/symmetry_detect.h"
+#include "theory/quantifiers/term_util.h"
+#include "theory/rewriter.h"
+
+using namespace std;
+
+namespace CVC4 {
+
+SymmetryDetect::Partition SymmetryDetect::detect(const vector<Node>& assertions)
+{
+  Node node = d_trueNode;
+  vector<Node>::const_iterator assertions_it = assertions.begin();
+
+  while (assertions_it != assertions.end())
+  {
+    node = NodeManager::currentNM()->mkNode(kind::AND, node, *assertions_it);
```

Comment:
Why create a node of the form `(and (and x y) z)` instead of `(and x y z)`? Also, what happens if one of the assertions contains a conjunction at the top level? Finally, loop can be rewritten as range-for.

Reply:
It does not matter whether it is (and (and x y) z) or (and x y z). If one of the assertions is a conjunction, we will recurse to collect all its children. That is why the function collectChildren is recursive. I do not see an easy way to write it non-recursively.
And I changed it to for loop.